File: adventofcode/2020/day15/day15.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

const (
	part1Limit = 2020
	part2Limit = 30000000
)

func parseInput(in string) []int32 {
	r := []int32{}
	ints := strings.Split(in, ",")
	for _, ini := range ints {
		i, _ := strconv.Atoi(ini)
		r = append(r, int32(i))
	}
	return r
}

func nextNumber(num, index int, numbers []int, ph, lh map[int]int) int {
	if index < len(numbers) {
		num = numbers[index]
		ph[num] = index
		return num
	}

	if n, ok := lh[num]; ok {
		nnum := n - ph[num]
		if _, ok = ph[nnum]; ok {
			lh[nnum] = index
		} else {
			ph[nnum] = index
		}
		ph[num] = index - 1
		num = nnum
	} else {
		num = 0
		lh[num] = index
	}
	return num
}

func part(nums []int, limit int) int {
	index := 0
	lh := make(map[int]int)
	ph := make(map[int]int)
	num := nums[0]
	for {
		num = nextNumber(num, index, nums, ph, lh)
		if index+1 == limit {
			return num
		}

		index++
	}
}

func fastPart(nums []int32, limit int32) string {
	var turn int32 = 0
	var num int32 = 0
	var lomemSize int32 = 10000000
	mem := map[int32]int32{}
	lomem := make([]int32, lomemSize)

	for ; turn < int32(len(nums)); turn++ {
		num = nums[turn]
		lomem[num] = turn + 1
	}

	for ; turn != limit; turn++ {
		var m int32
		var ok bool
		if num < lomemSize {
			m = lomem[num]
			ok = m != 0
			lomem[num] = turn
		} else {
			m, ok = mem[num]
			mem[num] = turn
		}

		if !ok {
			num = 0
		} else {
			num = turn - m
		}
	}

	return fmt.Sprint(num)
}

func main() {
	var input string
	fmt.Scanln(&input)
	nums := parseInput(input)
	fmt.Println(fastPart(nums, part1Limit))
	fmt.Println(fastPart(nums, part2Limit))
}

File: adventofcode/2020/day15/day15.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

PART1_LIMIT = 2020
PART2_LIMIT = 30_000_000

# ph : history of index for previous to last occurence of number
# lh : history of index for last occurence of number
def next_number(num, index, numbers, ph, lh)
  if index < numbers.size
    num = numbers[index]
    ph[num] = index
    return num
  end

  if lh[num] # second repeat
    nnum = lh[num] - ph[num]
    ph[nnum] ? lh[nnum] = index : ph[nnum] = index
    ph[num] = index - 1
    num = nnum
  else
    num = 0
    lh[num] = index # works because we know 0 has occurred before
  end
  num
end

def part(numbers, limit)
  index = 0
  lh = {}
  ph = {}
  num = numbers.first
  loop do
    num = next_number(num, index, numbers, ph, lh)
    return num if index + 1 == limit

    index += 1
  end
end

def fast_part(numbers, limit)
  lomem_size = 10_000_000
  mem = {}
  lomem = Array.new(lomem_size) { 0 }

  numbers.each.with_index { |n, i| lomem[n] = i + 1 }
  num = numbers[-1]
  turn = numbers.size

  while turn != limit
    m = -1
    ok = false
    if num < lomem_size
      m = lomem[num]
      ok = m != 0
      lomem[num] = turn
    else
      m = mem[num]
      ok = mem.key?(num)
      mem[num] = turn
    end
    num = if ok
            turn - m
          else
            0
          end
    turn += 1
  end
  num
end

start_numbers = read_input[0].split(',').map(&:to_i)
p fast_part(start_numbers, PART1_LIMIT)
p fast_part(start_numbers, PART2_LIMIT)
